shared/src/main/scala/mouse/boolean.scala

```diff
@@ -12,9 +12,16 @@ class ApplyIfPartiallyApplied[A](b: Boolean, a: A) {
   @inline def apply[B >: A](f: B => B): B = if (b) f(a) else a
 }
 
+class PureOrEmptyPartiallyApplied[F[_]](b: Boolean) {
+  @inline def apply[A](a: => A)(implicit F: Alternative[F]): F[A] =
+    if (b) F.pure(a) else F.empty
+}
+
 final class BooleanOps(private val b: Boolean) extends AnyVal {
 
-  @inline def option[A](a: => A): Option[A] = fold(Some(a), None)
```

Comment:
I think that the previous implementation is quite good. We don't need to change it — having fewer allocations is preferable.

Reply:
I'll make `PureOrEmptyPartiallyApplied` extend `AnyVal` to avoid an allocation in `pureOrEmpty`. Passing an implicit `Alternative[Option]` to the `apply` method won't cause an allocation either, because the `Alternative[Option]` instance in cats is a `val` that will be reused. So assuming the `AnyVal` change, there are zero allocations in this code, except for the unavoidable one when actually creating the result.

The previous implementation on the other hand uses `fold`, which takes two by-name parameters, and by-name parameters are compiled to function objects that need to be allocated, so you have two allocations right there.

I therefore cannot agree with your reasoning.

Reply:
I've now added the `AnyVal` optimization.

Reply:
Unfortunately, `AnyVal` is not so good, how it may seem. https://failex.blogspot.com/2017/04/the-high-cost-of-anyval-subclasses.html
Also, you can check similar syntax ops in `cats` https://github.com/typelevel/cats/blob/e0784fc0d95b9c4eed0bb1ddce6c69e3fd9a4a4f/core/src/main/scala/cats/syntax/option.scala#L372
Also, as Ben pointed, I don't sure if performance will not regress.

Reply:
And my initial point is not actually about `AnyVal` itself. I think we could have the new method, and it's quite good, but just I see no benefits in changing implementation here :)

Reply:
It's just common sense that when you add a generalization of an existing method, then you implement the more specific version in terms of the more general one. Sorry, I'm not reverting this change. Take it (like @benhutchison suggested) or leave it.